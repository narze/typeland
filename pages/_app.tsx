import Head from 'next/head'
import '../styles/base.css'
import { AuthProvider } from '../contexts/Auth'
import { StatsProvider } from '../contexts/Stats'
import { ThemeProvider, theme } from '@chakra-ui/core'
import React from 'react'

export default function MyApp({
  Component,
  pageProps,
}: {
  Component: any
  pageProps: any
}): JSX.Element {
  return (
    <React.StrictMode>
      <Head>
        <title>Typeland</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <AuthProvider>
        <StatsProvider>
          <ThemeProvider theme={theme}>
            <Component {...pageProps} />
          </ThemeProvider>
        </StatsProvider>
      </AuthProvider>
    </React.StrictMode>
  )
}
