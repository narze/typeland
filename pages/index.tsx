/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import tw from '@tailwindcssinjs/macro'
import { useEffect, useState } from 'react'
import { TypingArea } from '../components/TypingArea'
import { randomWords } from '../utils/wordsDb'
import { useStats } from '../contexts/Stats'
import { useAuth } from '@/contexts/Auth'
import Link from 'next/link'
import { auth as firebaseAuth } from '../config/firebase'

const s = {
  container: tw`
    h-screen min-w-full
    flex
    items-center justify-center
  `,
  main: [
    tw`
      max-w-1/2
    `,
    css`
      min-width: 640px;
    `,
  ],
  title: tw`
    text-4xl
    text-blue-600
    antialiased
  `,
  typingInput: tw`
    h-0
    w-0
  `,
  footer: tw`
    fixed
    bottom-0
    pb-8
    text-center
  `,
  header: tw`
    fixed
    top-0
    right-8
    pt-4
    text-right
  `,
  githubLink: tw`
    text-xs
  `,
  result: tw`
    mt-4
    text-center
    text-xl
  `,
  restart: tw`
    mt-2
    text-center
  `,
  restartButton: tw`
    py-2 px-4
    bg-transparent hover:bg-blue-500
    font-semibold
    hover:text-white
    border border-blue-500 hover:border-transparent
    rounded
  `,
  restartHint: tw`
    mt-1
    text-xs
  `,
  liveWpm: tw`
    text-base
  `,
}

export const Home = (): JSX.Element => {
  const [words, setWords] = useState([''])
  const [userTypeInput, setUserTypeInput] = useState([''])
  const [inputIsFocused, setInputIsFocused] = useState(true)
  const [finished, setFinished] = useState(false)
  const [started, setStarted] = useState(false)
  const [startTime, setStartTime] = useState(0)
  const [finishTime, setFinishTime] = useState(0)
  const [wpm, setWpm] = useState(0)
  const [liveWpm, setLiveWpm] = useState(0)
  const [elapsedMs, setElapsedMs] = useState(0)
  const [promptRestart, setPromptRestart] = useState(false)
  const [stats, dispatch] = useStats()
  const [auth] = useAuth()

  const DEFAULT_WORD_COUNT = 30
  const TIMER_LOOP_MS = 1000

  useEffect(() => {
    if (window._seed) {
      setWords(window._seed.words.split(' '))
    } else {
      setWords(randomWords(DEFAULT_WORD_COUNT))
    }
  }, [])

  useEffect(() => {
    let timer

    if (started) {
      timer = setInterval(() => {
        setElapsedMs((ms) => ms + TIMER_LOOP_MS)
      }, TIMER_LOOP_MS)
    } else {
      setElapsedMs(0)
      clearInterval(timer)
    }

    return () => clearInterval(timer)
  }, [started])

  useEffect(() => {
    setLiveWpm(Math.round((stats.correct * 60) / (elapsedMs / 1000.0)) || 0)
  }, [elapsedMs])

  const handleDelete = ({ word = false }) => {
    const newUserTypeInput = [...userTypeInput]

    if (word) {
      newUserTypeInput[newUserTypeInput.length - 1] = ''
    } else {
      newUserTypeInput[newUserTypeInput.length - 1] = newUserTypeInput[
        newUserTypeInput.length - 1
      ].slice(0, -1)
    }

    setUserTypeInput(newUserTypeInput)
  }

  const handleSpace = () => {
    if (!userTypeInput[userTypeInput.length - 1].length) {
      return
    }
    if (userTypeInput.length == words.length) {
      setFinished(true)
      return
    }

    const newUserTypeInput = [...userTypeInput]

    newUserTypeInput[newUserTypeInput.length] = ''

    setUserTypeInput(newUserTypeInput)
  }

  const handleEnter = () => {
    if (finished || promptRestart) {
      return restart()
    }

    if (started) {
      setPromptRestart(true)
    }
  }

  const handleType = (e) => {
    const { key, altKey, ctrlKey, metaKey } = e

    if (key == 'Backspace') {
      return handleDelete({ word: altKey || ctrlKey || metaKey })
    }

    if (key == 'Enter') {
      return handleEnter()
    }

    // Filter out modifiers
    if (key.length != 1 || altKey || ctrlKey || metaKey) {
      return
    }

    if (promptRestart) {
      setPromptRestart(false)
    }

    if (key == ' ') {
      return handleSpace()
    }

    if (!started) {
      setStarted(true)
    }

    const newUserTypeInput = [...userTypeInput]
    newUserTypeInput[newUserTypeInput.length - 1] = newUserTypeInput[
      newUserTypeInput.length - 1
    ].concat(key)
    setUserTypeInput(newUserTypeInput)

    if (userTypeInput.length == words.length) {
      if (
        newUserTypeInput[newUserTypeInput.length - 1] == words[words.length - 1]
      ) {
        setFinished(true)
      }
    }
  }

  const restart = () => {
    setWords(randomWords(DEFAULT_WORD_COUNT))
    setUserTypeInput([''])
    setInputIsFocused(true)
    setFinished(false)
    setStarted(false)
    setStartTime(0)
    setFinishTime(0)
    setWpm(0)
    setLiveWpm(0)
    setElapsedMs(0)
    setPromptRestart(false)
    dispatch({ type: 'reset' })
  }

  useEffect(() => {
    if (started && !finished) {
      setStartTime(+new Date())
    }

    if (started && finished) {
      setFinishTime(+new Date())
    }
  }, [started, finished])

  useEffect(() => {
    if (startTime && finishTime && !wpm) {
      setWpm(
        Math.round((stats.correct * 60) / ((finishTime - startTime) / 1000.0))
      )
    }
  }, [startTime, finishTime, wpm])

  return (
    <div
      css={s.container}
      onClick={() => document.getElementById('typingInput').focus()}
    >
      <div css={s.header}>
        {auth.user ? (
          <>
            <span>{auth.user.email}</span>
            <button
              css={tw`ml-4 cursor-pointer`}
              onClick={() => firebaseAuth.signOut()}
            >
              Logout
            </button>
          </>
        ) : (
          <Link href="/login">Login</Link>
        )}
      </div>

      <main css={s.main}>
        <h1 css={s.title}>Typeland</h1>

        <TypingArea
          words={words}
          userWords={userTypeInput}
          showCaret={inputIsFocused}
          finished={finished}
        />

        {started && !finished && <div css={s.liveWpm}>{liveWpm} wpm</div>}

        {started && (
          <div css={s.liveWpm}>
            Stats (correct/wrong/total) : {stats.correct}/{stats.wrong}/
            {stats.total}
          </div>
        )}

        {finished && finishTime && (
          <>
            <div css={s.result}>Good job! {wpm} wpm</div>
            <div css={s.restart}>
              <button css={s.restartButton} onClick={restart}>
                Restart
              </button>
              <div css={s.restartHint}>
                Click or press <code>Enter</code>
              </div>
            </div>
          </>
        )}

        {promptRestart && (
          <>
            <div css={s.restart}>
              <button css={s.restartButton} onClick={restart}>
                Restart
              </button>
              <div css={s.restartHint}>
                Press <code>Enter</code> again to restart typing.
              </div>
            </div>
          </>
        )}

        <input
          id="typingInput"
          data-testid="typingInput"
          css={s.typingInput}
          type="text"
          ref={(i) => i && i.focus()} // Autofocus
          onKeyDown={handleType}
          onBlur={() => {
            setInputIsFocused(false)
          }}
          onFocus={() => {
            setInputIsFocused(true)
          }}
          tabIndex={0}
        />
      </main>

      <div css={s.footer}>
        <a
          css={s.githubLink}
          href="https://github.com/narze/typeland"
          target="_blank"
          rel="noreferrer"
        >
          GitHub
        </a>
      </div>
    </div>
  )
}

export default Home
