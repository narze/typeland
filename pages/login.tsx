/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import tw from '@tailwindcssinjs/macro'
import { auth as firebaseAuth } from '../config/firebase'
import { useAuth } from '@/contexts/Auth'
import { useRouter } from 'next/router'

import {
  FormControl,
  FormLabel,
  Input,
  FormErrorMessage,
  Button,
} from '@chakra-ui/core'
import { useForm } from 'react-hook-form'
import { useState, useEffect } from 'react'

const s = {
  container: tw`
    h-screen min-w-full
    flex
    items-center justify-center
  `,
  main: [
    tw`
      max-w-1/2
    `,
    css`
      min-width: 640px;
    `,
  ],
  header: tw`
    fixed
    top-0
    right-8
    pt-4
    text-right
  `,
  title: tw`
    text-4xl
    text-blue-600
    antialiased
  `,
}

export const Login = (): JSX.Element => {
  const { handleSubmit, errors, register, formState } = useForm()
  const [auth] = useAuth()
  const [signupForm, setSignupForm] = useState(false)
  const router = useRouter()

  const signUp = ({ email, password }) => {
    return firebaseAuth
      .createUserWithEmailAndPassword(email, password)
      .then((response) => {
        // eslint-disable-next-line no-console
        console.log(response)
      })
  }

  const login = ({ email, password }) => {
    return firebaseAuth
      .signInWithEmailAndPassword(email, password)
      .then((response) => {
        // eslint-disable-next-line no-console
        console.log(response)
      })
  }

  function validateEmail(value) {
    let error
    if (!value) {
      error = 'Email is required'
    }
    return error || true
  }

  function validatePassword(value) {
    let error
    if (!value) {
      error = 'Password is required'
    } else if (value.length < 8) {
      error = 'Password must be longer than 8 characters'
    }
    return error || true
  }

  useEffect(() => {
    if (auth.user) {
      router.push('/')
    }
  }, [auth])

  if (signupForm) {
    return (
      <div css={s.container}>
        <div css={s.header}>
          <button css={tw`ml-4 cursor-pointer`} onClick={() => router.back()}>
            Back
          </button>
        </div>

        <main css={s.main}>
          <h1 css={s.title}>Sign Up</h1>

          <form onSubmit={handleSubmit(signUp)}>
            <FormControl isInvalid={errors.email}>
              <FormLabel htmlFor="email">Email</FormLabel>
              <Input
                id="email"
                name="email"
                ref={register({ validate: validateEmail })}
              />
              <FormErrorMessage>
                {errors.email && errors.email.message}
              </FormErrorMessage>
            </FormControl>

            <FormControl isInvalid={errors.password}>
              <FormLabel htmlFor="password">Password</FormLabel>
              <Input
                id="password"
                name="password"
                type="password"
                ref={register({ validate: validatePassword })}
              />
              <FormErrorMessage>
                {errors.password && errors.password.message}
              </FormErrorMessage>
            </FormControl>

            <div css={tw`mt-4`}>
              <Button
                variantColor="teal"
                isLoading={formState.isSubmitting}
                type="submit"
              >
                Submit
              </Button>

              <Button
                variant="link"
                ml={4}
                onClick={() => setSignupForm(false)}
              >
                Login
              </Button>
            </div>
          </form>
        </main>
      </div>
    )
  }

  return (
    <div css={s.container}>
      <div css={s.header}>
        <button css={tw`ml-4 cursor-pointer`} onClick={() => router.back()}>
          Back
        </button>
      </div>

      <main css={s.main}>
        <h1 css={s.title}>Login</h1>

        <form onSubmit={handleSubmit(login)}>
          <FormControl isInvalid={errors.email}>
            <FormLabel htmlFor="email">Email</FormLabel>
            <Input
              id="email"
              name="email"
              ref={register({ validate: validateEmail })}
            />
            <FormErrorMessage>
              {errors.email && errors.email.message}
            </FormErrorMessage>
          </FormControl>

          <FormControl isInvalid={errors.password}>
            <FormLabel htmlFor="password">Password</FormLabel>
            <Input
              id="password"
              name="password"
              type="password"
              ref={register({ validate: validatePassword })}
            />
            <FormErrorMessage>
              {errors.password && errors.password.message}
            </FormErrorMessage>
          </FormControl>

          <div css={tw`mt-4`}>
            <Button
              variantColor="teal"
              isLoading={formState.isSubmitting}
              type="submit"
            >
              Login
            </Button>

            <Button variant="link" ml={4} onClick={() => setSignupForm(true)}>
              Sign Up
            </Button>
          </div>
        </form>
      </main>
    </div>
  )
}

export default Login
