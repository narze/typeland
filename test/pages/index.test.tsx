import React from 'react'
import { render, fireEvent } from '../testUtils'
import userEvent from '@testing-library/user-event'
import { Home } from '../../pages/index'

describe('Home page', () => {
  it('renders title', () => {
    const { asFragment, getByText } = render(<Home />, {})
    getByText(/Typeland/i)

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders text to type', () => {
    const { getByText } = render(<Home />, {})

    getByText(/the quick brown fox jumps over the lazy dog/i)
  })

  it('highlights text which is typed', () => {
    const { asFragment, getByTestId, getAllByTestId } = render(<Home />, {})

    const typingInput = getByTestId('typingInput')

    fireEvent.keyDown(typingInput, { key: 't' })
    expect(getByTestId('correct')).toHaveTextContent('t')

    fireEvent.keyDown(typingInput, { key: 'h' })
    expect(getAllByTestId('correct')[0]).toHaveTextContent('t')
    expect(getAllByTestId('correct')[1]).toHaveTextContent('h')

    fireEvent.keyDown(typingInput, { key: 'x' })

    expect(getAllByTestId('correct')[0]).toHaveTextContent('t')
    expect(getAllByTestId('correct')[1]).toHaveTextContent('h')
    expect(getByTestId('wrong')).toHaveTextContent('x')

    expect(asFragment()).toMatchSnapshot()
  })

  it('can delete typed text with backspace', () => {
    const { asFragment, getByTestId, getAllByTestId, queryByTestId } = render(
      <Home />,
      {}
    )

    const typingInput = getByTestId('typingInput')

    fireEvent.keyDown(typingInput, { key: 't' })
    expect(getByTestId('correct')).toHaveTextContent('t')

    fireEvent.keyDown(typingInput, { key: 'h' })
    expect(getAllByTestId('correct')[0]).toHaveTextContent('t')
    expect(getAllByTestId('correct')[1]).toHaveTextContent('h')

    userEvent.type(typingInput, '{backspace}')
    expect(getByTestId('correct')).toHaveTextContent('t')

    fireEvent.keyDown(typingInput, { key: 'o' })
    expect(getByTestId('correct')).toHaveTextContent('t')
    expect(getByTestId('wrong')).toHaveTextContent('o')

    userEvent.type(typingInput, '{backspace}')
    expect(getByTestId('correct')).toHaveTextContent('t')
    expect(queryByTestId('wrong')).not.toBeInTheDocument()

    fireEvent.keyDown(typingInput, { key: 'h' })
    expect(getAllByTestId('correct')[0]).toHaveTextContent('t')
    expect(getAllByTestId('correct')[1]).toHaveTextContent('h')
    expect(queryByTestId('wrong')).not.toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot()
  })

  it('can delete whole incomplete word with option/ctrl/meta backspace', () => {
    pending()
  })

  it('blocks all modifiers except shift', () => {
    pending()
  })

  it('shows wpm when finished typing', () => {
    pending()
  })
})
