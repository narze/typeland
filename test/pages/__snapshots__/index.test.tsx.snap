// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page can delete typed text with backspace 1`] = `
<DocumentFragment>
  .emotion-8 {
  height: 100vh;
  min-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  --text-opacity: 1;
  font-size: 2.25rem;
  color: #1c64f2;
  color: rgba(28,100,242,var(--text-opacity));
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.emotion-4 {
  font-size: 1.25rem;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.emotion-3 {
  --border-opacity: 1;
  height: 1rem;
  border-width: 1px;
  border-color: #3f83f8;
  border-color: rgba(63,131,248,var(--border-opacity));
  width: 1px;
}

.emotion-5 {
  height: 0;
  width: 0;
}

.emotion-7 {
  position: fixed;
  bottom: 0;
  padding-bottom: 2rem;
  text-align: center;
}

.emotion-6 {
  font-size: 0.75rem;
}

.emotion-1 {
  --text-opacity: 1;
  color: #057a55;
  color: rgba(5,122,85,var(--text-opacity));
}

<div
    class="emotion-8"
  >
    <main>
      <h1
        class="emotion-0"
      >
        Typeland
      </h1>
      <p
        class="emotion-4"
      >
        the quick brown fox jumps over the lazy dog
        <br />
        <span>
          <span
            class="correct emotion-1"
            data-testid="correct"
          >
            t
          </span>
          <span
            class="correct emotion-1"
            data-testid="correct"
          >
            h
          </span>
          <span
            class="emotion-3"
            data-testid="caret"
          />
        </span>
        <span>
           
        </span>
      </p>
      <input
        class="emotion-5"
        data-testid="typingInput"
        id="typingInput"
        tabindex="0"
        type="text"
      />
    </main>
    <div
      class="emotion-7"
    >
      <a
        class="emotion-6"
        href="https://github.com/narze/typeland"
        rel="noreferrer"
        target="_blank"
      >
        GitHub
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Home page highlights text which is typed 1`] = `
<DocumentFragment>
  .emotion-9 {
  height: 100vh;
  min-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  --text-opacity: 1;
  font-size: 2.25rem;
  color: #1c64f2;
  color: rgba(28,100,242,var(--text-opacity));
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.emotion-5 {
  font-size: 1.25rem;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.emotion-4 {
  --border-opacity: 1;
  height: 1rem;
  border-width: 1px;
  border-color: #3f83f8;
  border-color: rgba(63,131,248,var(--border-opacity));
  width: 1px;
}

.emotion-6 {
  height: 0;
  width: 0;
}

.emotion-8 {
  position: fixed;
  bottom: 0;
  padding-bottom: 2rem;
  text-align: center;
}

.emotion-7 {
  font-size: 0.75rem;
}

.emotion-1 {
  --text-opacity: 1;
  color: #057a55;
  color: rgba(5,122,85,var(--text-opacity));
}

.emotion-3 {
  --text-opacity: 1;
  color: #f98080;
  color: rgba(249,128,128,var(--text-opacity));
}

<div
    class="emotion-9"
  >
    <main>
      <h1
        class="emotion-0"
      >
        Typeland
      </h1>
      <p
        class="emotion-5"
      >
        the quick brown fox jumps over the lazy dog
        <br />
        <span>
          <span
            class="correct emotion-1"
            data-testid="correct"
          >
            t
          </span>
          <span
            class="correct emotion-1"
            data-testid="correct"
          >
            h
          </span>
          <span
            class="wrong emotion-3"
            data-testid="wrong"
          >
            x
          </span>
          <span
            class="emotion-4"
            data-testid="caret"
          />
        </span>
        <span>
           
        </span>
      </p>
      <input
        class="emotion-6"
        data-testid="typingInput"
        id="typingInput"
        tabindex="0"
        type="text"
      />
    </main>
    <div
      class="emotion-8"
    >
      <a
        class="emotion-7"
        href="https://github.com/narze/typeland"
        rel="noreferrer"
        target="_blank"
      >
        GitHub
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Home page renders title 1`] = `
<DocumentFragment>
  .emotion-6 {
  height: 100vh;
  min-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  --text-opacity: 1;
  font-size: 2.25rem;
  color: #1c64f2;
  color: rgba(28,100,242,var(--text-opacity));
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.emotion-2 {
  font-size: 1.25rem;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.emotion-1 {
  --border-opacity: 1;
  height: 1rem;
  border-width: 1px;
  border-color: #3f83f8;
  border-color: rgba(63,131,248,var(--border-opacity));
  width: 1px;
}

.emotion-3 {
  height: 0;
  width: 0;
}

.emotion-5 {
  position: fixed;
  bottom: 0;
  padding-bottom: 2rem;
  text-align: center;
}

.emotion-4 {
  font-size: 0.75rem;
}

<div
    class="emotion-6"
  >
    <main>
      <h1
        class="emotion-0"
      >
        Typeland
      </h1>
      <p
        class="emotion-2"
      >
        the quick brown fox jumps over the lazy dog
        <br />
        <span>
          <span
            class="emotion-1"
            data-testid="caret"
          />
        </span>
        <span>
           
        </span>
      </p>
      <input
        class="emotion-3"
        data-testid="typingInput"
        id="typingInput"
        tabindex="0"
        type="text"
      />
    </main>
    <div
      class="emotion-5"
    >
      <a
        class="emotion-4"
        href="https://github.com/narze/typeland"
        rel="noreferrer"
        target="_blank"
      >
        GitHub
      </a>
    </div>
  </div>
</DocumentFragment>
`;
